src/common/type/pack.h

```diff
@@ -0,0 +1,500 @@
+/***********************************************************************************************************************************
+Pack Type
+
+The pack type encodes binary data compactly while still maintaining structure and strict typing. The idea is based on Thrift,
+ProtocolBuffers, and Avro, compared here: https://medium.com/better-programming/use-binary-encoding-instead-of-json-dec745ec09b6.
+The pack type has been further optimized to balance between purely in-memory structures and those intended to be passed via a
+protocol or saved in a file.
+
+Integers are stored with base-128 varint encoding which is equivalent to network byte order, i.e., the endianness of the sending and
+receiving host don't matter.
+
+The overall idea is similar to JSON but IDs are used instead of names, typing is more granular, and the representation is far more
+compact. A pack can readily be converted to JSON but the reverse is not as precise due to loose typing in JSON. A pack is a stream
+format, i.e. it is intended to be read in order from beginning to end.
+
+Fields in a pack are identified by IDs. A field ID is stored as a delta from the previous ID, which is very efficient, but means
+that reading from the middle is generally not practical. The size of the gap between field IDs is important -- a gap of 1 never
+incurs extra cost, but depending on the field type larger gaps may require additional bytes to store the field ID delta.
+
+NULLs are not stored in a pack and are therefore not typed. A NULL is essentially just a gap in the field IDs. Fields that are
+frequently NULL are best stored at the end of an object.
+
+A pack is an object by default. Objects can store fields, objects, or arrays. Objects and arrays will be referred to collectively as
+containers. Fields contain data to be stored, e.g. integers, strings, etc.
+
+Here is a simple example of a pack:
+
+PackWrite *write = pckWriteNew(buffer);
+pckWriteU64P(write, 77);
+pckWriteBoolP(write, false, .defaultNull = true);
+pckWriteI32P(write, -1, .defaultNull = true, .defaultValue = -1);
+pckWriteStringP(write, STRDEF("sample"));
+pckWriteEndP();
+
+A string representation of this pack is `1:uint64:77,4:str:sample`. The boolean was not stored because we requested NULL on default
+and the default was false (because it was not set to anything else). The int32 field was also not stored but we set an explicit
+default value. Note that there is a gap in the ID stream, which represents the NULL values.
+
+This pack can be read with:
+
+PackRead *read = pckReadNew(buffer);
+pckReadU64P(read);
+pckReadBoolP(read, .defaultNull = true);
+pckReadI32P(read, .defaultNull = true, .defaultValue = -1);
```

Comment:
We're reading so why do we have to specify the defaultValue again?  I see in line 48 you state "If we don't care about the NULLs" then we don't need to specify defaultNull and then defaultValue also because - what - defaulValue is treated as NULL?

Reply:
Clarified in 021a830254e62.